pkg/kubelet/cm/devicemanager/manager_test.go

```diff
@@ -753,3 +761,108 @@ func TestSanitizeNodeAllocatable(t *testing.T) {
 	// allocatable in nodeInfo is more than needed, should skip updating
 	as.Equal(2, int(allocatableScalarResources[v1.ResourceName(resourceName2)]))
 }
+
+func TestDevicePreStart(t *testing.T) {
+	// Ensures that if device manager is indicated to invoke `PreStart` RPC
+	// by device plugin, then device manager invokes preStart at endpoint interface.
+	// Also verifies that final allocation of mounts, envs etc is same as expected.
+	res1 := TestResource{
+		resourceName:     "domain1.com/resource1",
+		resourceQuantity: *resource.NewQuantity(int64(2), resource.DecimalSI),
+		devs:             []string{"dev1", "dev2"},
+	}
+	as := require.New(t)
+	podsStub := activePodsStub{
+		activePods: []*v1.Pod{},
+	}
+	tmpDir, err := ioutil.TempDir("", "checkpoint")
+	as.Nil(err)
+	defer os.RemoveAll(tmpDir)
+	nodeInfo := getTestNodeInfo(v1.ResourceList{})
+	monitorCallback := func(resourceName string, added, updated, deleted []pluginapi.Device) {}
+	testManager := &ManagerImpl{
+		socketdir:        tmpDir,
+		callback:         monitorCallback,
+		healthyDevices:   make(map[string]sets.String),
+		allocatedDevices: make(map[string]sets.String),
+		endpoints:        make(map[string]endpoint),
+		pluginOpts:       make(map[string]*pluginapi.DevicePluginOptions),
+		podDevices:       make(podDevices),
+		activePods:       podsStub.getActivePods,
+		sourcesReady:     &sourcesReadyStub{},
+	}
+	testManager.store, _ = utilstore.NewFileStore("/tmp/", utilfs.DefaultFs{})
+	testManager.healthyDevices[res1.resourceName] = sets.NewString()
+	testManager.pluginOpts[res1.resourceName] = &pluginapi.DevicePluginOptions{PreStartRequired: true}
+	for _, dev := range res1.devs {
+		testManager.healthyDevices[res1.resourceName].Insert(dev)
+	}
+	ch := make(chan []string, 1)
+	testManager.endpoints[res1.resourceName] = &MockEndpoint{
+		initChan: ch,
+		allocateFunc: func(devs []string) (*pluginapi.AllocateResponse, error) {
```

Comment:
This pattern is now present twice in the code.
Can we have a function call replacing this ?

Something like:
```golang
newAllocateStubFunc(map[string]string{
	"dev1": newAllocateResp(
		[]*pluginapi.DeviceSpec{
			*pluginapi.DeviceSpec{ContainerPath: "/dev/aaa", HostPath: "/dev/aaa", Permissions: "mrw"},
			*pluginapi.DeviceSpec{ContainerPath: "/dev/bbb", HostPath: "/dev/bbb", Permissions: "mrw"},
		},
		[]*pluginapi.Mount{
			*pluginapi.Mount{ContainerPath: "/container_dir1/file1", HostPath: "host_dir1/file1", ReadOnly: true}
		},
		map[string]string{
			"key1": "val1"
		},
	),
//....
})
```